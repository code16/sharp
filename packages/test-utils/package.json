{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}