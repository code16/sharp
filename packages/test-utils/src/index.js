export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}