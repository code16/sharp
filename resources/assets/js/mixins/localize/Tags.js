import localizeSelect from './Select';
import { localeObject } from "./utils";
import { lang } from "../../index";

export default {
    extends: localizeSelect,
    methods: {
        localizeLabel(label) { // display
            return this.isLocalized ? label[this.locale] || lang('form.tags.unknown_label') : label;
        },
        localizedTagLabel(text) { // data
            return this.isLocalized ? localeObject({
                locales:this.locales,
                resolve:l => l===this.locale ? text : null
            }) : text;
        },
        localizedCustomLabel(option) {
            return this.localizeLabel(option.label)
        }
    }
}