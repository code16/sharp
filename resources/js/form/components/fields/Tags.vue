<script setup lang="ts">
    import { FormFieldEmits, FormFieldProps } from "@/form/types";
    import { FormTagsFieldData } from "@/types";
    import { TagsInput, TagsInputInput, TagsInputItem, TagsInputItemText, TagsInputItemDelete } from "@/components/ui/tags-input";
    import { ComboboxAnchor, ComboboxContent, ComboboxPortal, ComboboxRoot, ComboboxInput } from "reka-ui";
    import { computed, ref } from "vue";
    import { __ } from "@/utils/i18n";
    import { CommandEmpty, CommandGroup, CommandItem, CommandList, CommandSeparator } from "@/components/ui/command";
    import FormFieldLayout from "@/form/components/FormFieldLayout.vue";
    import { useFullTextSearch } from "@/composables/useFullTextSearch";

    const props = defineProps<FormFieldProps<FormTagsFieldData>>();
    const emit = defineEmits<FormFieldEmits<FormTagsFieldData>>();

    const open = ref(false);
    const searchTerm = ref('');

    let itemKeyIndex = 0;
    const itemKey = Symbol('itemKey');
    function withItemKey(option: FormTagsFieldData['options'][0]) {
        return {
            ...option,
            [itemKey as any]: itemKeyIndex++,
        }
    }

    function onCreateClick() {
        emit('input', [
            ...(props.value ?? []),
            withItemKey({
                id: null,
                label: searchTerm.value,
            }),
        ]);
    }

    function onOptionClick(option: FormTagsFieldData['options'][0]) {
        emit('input', [
            ...(props.value ?? []),
            withItemKey(option),
        ]);
    }

    function onDeleteClick(item: FormTagsFieldData['value'][0]) {
        emit('input', props.value.filter(i => i[itemKey] !== item[itemKey]));
    }

    const { fullTextSearch } = useFullTextSearch(() => props.field.options, { id: 'id', searchKeys: ['label'] });
    const filteredOptions = computed(() => {
        const filtered = searchTerm.value.length > 0
            ? fullTextSearch(searchTerm.value)
            : props.field.options;
        return filtered
            .filter(o => !props.value?.find(v => o.id != null && v.id != null && o.id === v.id)); // show only unselected options
    });
    const canCreate = computed(() => {
        return props.field.creatable
            && searchTerm.value.length > 0
            && !props.value?.find(v => v.label === searchTerm.value)
            && !props.field.options.find(o => o.label === searchTerm.value);
    });

    emit('input', props.value?.map(option => withItemKey(option)), { skipRefresh: true });
</script>

<template>
    <FormFieldLayout v-bind="props" field-group v-slot="{ id, ariaDescribedBy }">
        <ComboboxRoot
            class="flex-1"
            :model-value="props.value"
            v-model:open="open"
            ignore-filter
        >
            <ComboboxAnchor>
                <TagsInput
                    class="ring-offset-background data-disabled:pointer-events-none data-disabled:opacity-50 has-[:focus-visible]:ring-ring has-focus-visible:ring-2 has-focus-visible:ring-offset-2"
                    :model-value="props.value"
                    :display-value="(item: typeof props.value[0]) => item.label ?? item.id"
                    :disabled="props.field.readOnly"
                    @click="open = true; $refs.input.$el.focus()"
                >
                    <template v-for="item in value" :key="item[itemKey]">
                        <TagsInputItem :value="item">
                            <TagsInputItemText />
                            <TagsInputItemDelete
                                :aria-labelledby="null"
                                :aria-label="__('sharp::form.tags.tag_delete_button.aria_label', { option_label: item.label })"
                                @click.stop="onDeleteClick(item)"
                            />
                        </TagsInputItem>
                    </template>

                    <ComboboxInput v-model="searchTerm" :placeholder="props.field.placeholder ?? __('sharp::form.multiselect.placeholder')" as-child>
                        <TagsInputInput :id="id" :aria-describedby="ariaDescribedBy" :disabled="props.field.readOnly" class="flex-1 w-40" autocomplete="off" @keydown.enter.prevent ref="input" />
                    </ComboboxInput>
                </TagsInput>
            </ComboboxAnchor>
            <ComboboxPortal>
                <CommandList
                    position="popper"
                    position-strategy="absolute"
                    :avoid-collisions="false"
                    class="z-50 w-(--reka-popper-anchor-width) rounded-md mt-2 border bg-popover text-popover-foreground shadow-md outline-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2"
                >
                    <CommandEmpty>
                        {{ __('sharp::form.autocomplete.no_results_text') }}
                    </CommandEmpty>
                    <template v-if="canCreate">
                        <CommandGroup>
                            <CommandItem :value="{ id: null, label: searchTerm }" @select.prevent="onCreateClick()" :key="searchTerm">
                                {{ props.field.createText }} “{{ searchTerm }}”
                            </CommandItem>
                        </CommandGroup>
                        <CommandSeparator class="last:hidden" />
                    </template>

                    <template v-if="filteredOptions.length">
                        <CommandGroup>
                            <template
                                v-for="option in filteredOptions"
                                :key="option.id"
                            >
                                <CommandItem
                                    :value="option"
                                    @select.prevent="onOptionClick(option)"
                                >
                                    {{ option.label }}
                                </CommandItem>
                            </template>
                        </CommandGroup>
                    </template>
                </CommandList>
            </ComboboxPortal>
        </ComboboxRoot>
    </FormFieldLayout>
</template>
