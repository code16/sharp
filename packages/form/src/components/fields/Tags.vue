<template>
    <Multiselect
        class="SharpTags"
        :value="tags"
        :options="indexedOptions"
        :placeholder="dynamicPlaceholder"
        :tag-placeholder="createText"
        :max="maxTagCount"
        :taggable="creatable"
        :close-on-select="false"
        :disabled="readOnly"
        track-by="_internalId"
        label="label"
        :custom-label="localizedCustomLabel"
        multiple searchable hide-selected
        :show-labels="false"
        @search-change="handleTextInput"
        @input="handleInput"
        @tag="handleNewTag"
        ref="multiselect"
    />
</template>

<script>
    import { Multiselect } from 'sharp-ui';
    import localize from '../../mixins/localize/Tags';

    class LabelledItem {
        constructor(item) {
            this.id = item.id;
            this.label = item.label;
        }

        set internalId(id) {
            this._internalId = id;
        }

        get internalId() {
            return this._internalId;
        }
    }

    class Option extends LabelledItem {
    }
    class Tag extends LabelledItem {
    }

    export default {
        name: 'SharpTags',
        mixins:[localize],
        components: {
            Multiselect
        },
        props: {
            value: Array, // [{id:0, label: 'AAA'}, ...]
            options: Array, // [{id:0, label:'AAA'}, ...]
            placeholder: String,
            maxTagCount: Number,
            createText: String,
            creatable: {
                type: Boolean,
                default: true
            },
            readOnly:Boolean,
        },
        data() {
            return {
                tags: [],
                lastIndex: 0
            }
        },
        computed: {
            indexedOptions() {
                return this.options.map(this.patchOption);
            },
            dynamicPlaceholder() {
                return this.tags.length < (this.maxTagCount || Infinity) ? this.placeholder : "";
            },
            ids() {
                return this.tags.map(t=>t.internalId);
            }
        },
        watch: {
            tags: 'onTagsChanged'
        },
        methods: {
            patchOption(option, index) {
                let patchedOption = new Option(option);
                patchedOption.internalId = index;
                return patchedOption;
            },
            patchTag(tag) {
                let matchedOption = this.indexedOptions.find(o => o.id === tag.id);
                let patchedTag = new Tag(matchedOption);
                patchedTag.internalId = matchedOption.internalId;
                return patchedTag;
            },
            handleNewTag(val) {
                let newTag = new Tag({id: null, label: this.localizedTagLabel(val) });
                newTag.internalId = this.lastIndex++;
                this.tags.push(newTag);
            },
            handleInput(val) {
                this.tags = val;
            },
            handleTextInput(text) {
                if (text.length > 0 && this.$refs.multiselect.filteredOptions.length > 1) {
                    this.$refs.multiselect.pointer = 1;
                }
                else this.$refs.multiselect.pointer = 0
            },
            onTagsChanged() {
                this.$emit('input', this.tags.map(t => new Tag(t)));
            }
        },
        created() {
            this.lastIndex += this.options.length;
            this.tags = (this.value||[]).map(this.patchTag);
        }
    }
</script>