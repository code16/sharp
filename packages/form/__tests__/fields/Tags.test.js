import Vue from 'vue';
import Tags from '../../src/components/fields/Tags.vue';

import { MockI18n } from '@sharp/test-utils';

describe('tags-field', () => {
    Vue.use(MockI18n);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-tags :value="value"
                            placeholder="placeholder"
                            :options="[{id:3, label:'AAA'}, {id:4, label:'BBB'}]" 
                            :max-tag-count="2"
                            :read-only="readOnly"
                            create-text="createText"
                            creatable
                            @input="inputEmitted($event)">
                </sharp-tags>
            </div>
        `;
        MockI18n.mockLangFunction();
    });

    test('can mout Tags field', async () => {
        await createVm();
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('expose appropriate props to multiselect component', async () => {
        let $tags = await createVm({
            propsData: {
                readOnly: true,
                value: [{ id:3, label:'AAA'}]
            },
        });

        let { multiselect } = $tags.$refs;

        expect(multiselect.$props).toMatchObject({
            value: [{id:3, label:'AAA'}],
            placeholder: 'placeholder',
            tagPlaceholder: 'createText',
            taggable: true,
            max: 2,
            closeOnSelect: false,
            multiple: true,
            searchable: true,
            hideSelected: true,
            disabled: true
        });
    });

    test('instantiate proper internal ids', async () => {
        let $tags = await createVm({
            propsData: {
                value: [{ id:4, label:'BBB'}]
            },
        });

        expect($tags.tags[0].internalId).toEqual(1); // ( second option internal id )

        expect($tags.indexedOptions[0].internalId).toEqual(0);
        expect($tags.indexedOptions[1].internalId).toEqual(1);
    });

    test('adding a tag & with proper internal id', async () => {
        let $tags = await createVm({
            propsData: {
                value: []
            },
        });

        let { multiselect } = $tags.$refs;

        expect($tags.tags.length).toBe(0);

        multiselect.$emit('tag', 'CCC');

        await Vue.nextTick();

        expect($tags.tags.length).toBe(1);
        expect($tags.tags[0]).toEqual({ id:null, label:'CCC', _internalId: 2 });
    });

    test('emit a tag properly', async () => {
        let $tags = await createVm({
            propsData: {
                value: [{ id:4, label:'BBB'}]
            },
        });

        let { multiselect } = $tags.$refs;
        let inputEmitted = jest.fn();

        multiselect.$emit('tag', 'CCC');

        $tags.$on('input', inputEmitted);
        await Vue.nextTick();

        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith([{ id:4, label:'BBB'},{ id:null, label:'CCC' }]);
    });

});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],

        components: {
            'sharp-tags': Tags
        },

        props:['value','readOnly'],

        'extends':{
           methods: {
               inputEmitted:()=>{}
           }
        },
    });

    await Vue.nextTick();

    return vm.$children[0];
}