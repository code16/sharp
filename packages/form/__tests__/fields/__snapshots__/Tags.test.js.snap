// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tags-field can mout Tags field 1`] = `
<div id="app">
	<div tabindex="-1" class="multiselect SharpTags">
		<svg width="10" height="5" viewBox="0 0 10 5" fill-rule="evenodd" class="multiselect__select">
			<path d="M10 0L5 5 0 0z"></path>
		</svg>
		<div class="multiselect__tags">
			<div class="multiselect__tags-wrap" style="display: none;"></div>
			<!---->
			<div class="multiselect__spinner" style="display: none;"></div>
			<input name="" type="text" autocomplete="off" placeholder="placeholder" tabindex="0" class="multiselect__input" style="width: 0px; position: absolute; padding: 0px;">
			<!---->
			<span class="multiselect__placeholder">
				placeholder
			</span>
		</div>
		<div tabindex="-1" class="multiselect__content-wrapper" style="max-height: 300px; display: none;">
			<ul class="multiselect__content" style="display: inline-block;">
				<!---->
				<li class="multiselect__element">
					<span data-select="" data-selected="" data-deselect="" class="multiselect__option">
						<span>AAA</span>
					</span>
					<!---->
				</li>
				<li class="multiselect__element">
					<span data-select="" data-selected="" data-deselect="" class="multiselect__option">
						<span>BBB</span>
					</span>
					<!---->
				</li>
				<li style="display: none;">
					<span class="multiselect__option">No elements found. Consider changing the search query.</span>
				</li>
				<li style="display: none;">
					<span class="multiselect__option">List is empty.</span>
				</li>
			</ul>
		</div>
	</div>
</div>
`;
